File: android/rctmgl/src/main/java-v10/com/mapbox/rctmgl/components/mapview/helpers/CameraChangeTracker.java
package com.mapbox.rctmgl.components.mapview.helpers;

import android.util.Log;

public class CameraChangeTracker {
    public static final int USER_GESTURE = 1;
    public static final int DEVELOPER_ANIMATION = 2;
    public static final int SDK_ANIMATION = 3;
    public static final int EMPTY = -1;

    private int reason = EMPTY;
    private boolean isAnimating;

    public void setReason(int reason) {
        this.reason = reason;
    }

    public void setIsAnimating(boolean isAnimating) {
        this.isAnimating = isAnimating;
    }

    public boolean isUserInteraction() {
        return reason == USER_GESTURE || reason == DEVELOPER_ANIMATION;
    }

    public boolean isAnimated() {
        return reason == DEVELOPER_ANIMATION || reason == SDK_ANIMATION;
    }

    public boolean isAnimating() {
        return this.isAnimating;
    }

    public boolean isEmpty() {
        return reason == EMPTY;
    }
}

File: android/src/main/java/com/rnmapbox/rnmbx/components/mapview/helpers/CameraChangeTracker.kt
package com.rnmapbox.rnmbx.components.mapview.helpers

import com.rnmapbox.rnmbx.components.mapview.MapGestureType

enum class CameraChangeReason {
    USER_GESTURE,
    DEVELOPER_ANIMATION,
    SDK_ANIMATION
}

class CameraChangeTracker {
    private val reasonByGesture = mutableMapOf<MapGestureType, CameraChangeReason>()
    var isAnimating = false

    fun setReason(type: MapGestureType, reason: CameraChangeReason) {
        reasonByGesture[type] = reason
    }

    fun clear() {
        reasonByGesture.clear()
    }

    fun clearReason(type: MapGestureType) {
        reasonByGesture.remove(type)
    }

    val isUserInteraction: Boolean
        get() = reasonByGesture.values.any { it == CameraChangeReason.USER_GESTURE }
    val isAnimated: Boolean
        get() = reasonByGesture.values.any { it == CameraChangeReason.DEVELOPER_ANIMATION || it == CameraChangeReason.SDK_ANIMATION }
    val isEmpty: Boolean
        get() = reasonByGesture.isEmpty()
}
